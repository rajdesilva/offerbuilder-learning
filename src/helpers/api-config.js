import { init as initApm } from "@elastic/apm-rum";

export const baseURL = {
  API_URL_BASE: process.env.REACT_APP_BASE_API_URL,
  LOGIN_URL: process.env.REACT_APP_LOGIN_URL,
  LOGOUT_URL: process.env.REACT_APP_LOGOUT_URL,
  PROFILE_URL: process.env.REACT_APP_USER_PROFILE_URL,
  HISTORY_URL: process.env.REACT_APP_USER_HISTORY_URL,
};

export const apm = initApm({
  // Set required service name (allowed characters: a-z, A-Z, 0-9, -, _, and space)
  serviceName: process.env.REACT_APP_APM_SERVICE_NAME,
  serverUrl: process.env.REACT_APP_APM_SERVICE_URL,

  // Set service version (required for sourcemap feature)
  serviceVersion: process.env.REACT_APP_APM_SERVICE_VERSION,
});