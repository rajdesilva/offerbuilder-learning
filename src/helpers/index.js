export * from './api-config';
export * from './history';
