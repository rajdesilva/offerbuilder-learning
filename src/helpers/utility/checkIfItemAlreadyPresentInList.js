export const checkIfItemAlreadyPresentInList = (
  itemToCheck,
  itemList
) => {
  if(itemList && itemToCheck && itemToCheck.id) {
    return (
      itemList.findIndex(
        (item) => item.id === itemToCheck.id
      ) !== -1
    );
  }
  return false;
};
