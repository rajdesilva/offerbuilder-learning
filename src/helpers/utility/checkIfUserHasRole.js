import { store } from "../../redux/store";

export const checkIfUserHasRole = (roleToFind) => {
  try {
    const userRoles =
      window.getValue(store.getState(), "userinfo.userDetails.roles") || [];
    return userRoles.findIndex((role) => role === roleToFind) !== -1;
  } catch (error) {
    console.log("error = ", error.toString());
  }
  return false;
};
