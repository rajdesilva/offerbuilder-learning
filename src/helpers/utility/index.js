export { createSearchQueryURL } from './createSearchQueryURL';
export { isParentOrChildrenPropertyAdded } from './isParentOrChildrenPropertyAdded';
export { isPropertySelected } from './isPropertySelected';
export { mapBrands } from './mapBrands';
export { mapOnlySuppliers } from './mapOnlySuppliers';
export { parseFormDescriptionsAndImagesForOffer } from './parseFormDescriptionsAndImagesForOffer';
export { removePropertyForSelectedChannels } from './removePropertyForSelectedChannels';
export { removePropertyForSelectedSuppliers } from './removePropertyForSelectedSuppliers';
export { getParsedProperty } from './getParsedProperty';
export { getFormattedValue } from './getFormattedValue';
export { getValue } from './getValue';
export { parsePropertyMarketingInfoToFormData } from './parsePropertyMarketingInfoToFormData';
export { createOfferRequestObject } from './createOfferRequestObject';
export { getParsedOffers } from './getParsedOffers';
export { getListOfPropertiesForSupplier } from './getListOfPropertiesForSupplier';
export { getListOfPropertiesForChannels } from './getListOfPropertiesForChannels';
export { getInitializedState } from './getInitializedState';
export { getParsedNewOfferSettingsObject } from './getParsedNewOfferSettingsObject';
export { getListOfPropertiesForStorefront } from './getListOfPropertiesForStorefront';
export { checkIfMultipleChannelPresentForTarget } from './checkIfMultipleChannelPresentForTarget';
export { checkIfMultipleSupplierPresentForTarget } from './checkIfMultipleSupplierPresentForTarget';
export { checkIfMultipleStorefrontPresentForTarget } from './checkIfMultipleStorefrontPresentForTarget';
export { removePropertyForSelectedStorefront } from './removePropertyForSelectedStorefront';
export { checkIfMultipleBrandPresentForTarget } from './checkIfMultipleBrandPresentForTarget';
export { getListOfPropertiesForBrand } from './getListOfPropertiesForBrand';
export { getSearchStateFromRedux } from './getSearchStateFromRedux';
export { getObjectStringify } from './getObjectStringify';
export { parseOfferMarketingFromOffer } from './parseOfferMarketingFromOffer';
export { parseOfferMarketingLanguagesFromOffer } from './parseOfferMarketingLanguagesFromOffer';
export { getFilterData } from './getFilterData';
export { getSearchAndFilterQueryURL } from './getSearchAndFilterQueryURL';
export { getSearchQueryURLForEditAddProperties } from "./getSearchQueryURLForEditAddProperties";
export { isSavedOfferProperty } from './isSavedOfferProperty';
export { getParsedPropertiesForDeeplinkSettings } from "./getParsedPropertiesForDeeplinkSettings";
export { resetApplicationReduxAndStorage } from "./resetApplicationReduxAndStorage";
export { updateSettingsBrandWithSearchedValues } from "./updateSettingsBrandWithSearchedValues";
export { checkIfUserHasRole } from "./checkIfUserHasRole";
export { getUserOfferBuilderRole } from "./getUserOfferBuilderRole";
export { checkIfNewPropertiesAdded } from "./checkIfNewPropertiesAdded";
export { getAllBrandsWithoutKeysFromTempBrands } from "./getAllBrandsWithoutKeysFromTempBrands";
export { checkIfItemAlreadyPresentInList } from "./checkIfItemAlreadyPresentInList";
export { checkIfTravellingDateValidAsPerBookingDate } from "./checkIfTravellingDateValidAsPerBookingDate";
export { checkIfPropertiesHaveMultipleLocations } from "./checkIfPropertiesHaveMultipleLocations";
export { displayMargin } from "./displayMargin";
export { getFilterCount } from "./getFilterCount";
export { labelForBrands } from "./labelForBrands";
export { getDistanceFromLatLonInKm } from "./getDistanceFromLatLonInKm";
export { getTrustYouIdListFromProperties } from "./getTrustYouIdListFromProperties";
export { getSearchParamsToFetchMarketPrice } from "./getSearchParamsToFetchMarketPrice";
export { getMarketPriceForPropertyFromList } from "./getMarketPriceForPropertyFromList";
export { getImageFilterCount } from "./getImageFilterCount";
export { getSearchAndFilterImagesQueryUrl } from "./getSearchAndFilterImagesQueryUrl";
export { getImageName } from "./getImageName";
export { updateReduxWithOfferData } from "./updateReduxWithOfferData";
export { getOfferTypeFromPropertiesSelected } from "./getOfferTypeFromPropertiesSelected";