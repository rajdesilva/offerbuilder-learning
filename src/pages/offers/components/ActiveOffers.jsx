import React, { Fragment, useState, useCallback, useMemo } from "react";
import {
  Col,
  Row,
  Empty,
  Button,
  Skeleton,
  Menu,
  Dropdown,
  Divider,
  Table,
} from "antd";
import { useDispatch, useSelector } from "react-redux";
import moment from "moment/min/moment-with-locales";
import defaultImg from "../../../assets/images/default-img.svg";
import { FormattedMessage } from "react-intl";
import { EditFilled, MoreOutlined } from "@ant-design/icons";
import { searchAndFilterOffer } from "../service";
import {
  checkIfUserHasRole,
  getParsedOffers,
  parseOfferMarketingFromOffer,
  parseOfferMarketingLanguagesFromOffer,
} from "../../../helpers/utility";
import ChangeOfferStatusModal from "./ChangeOfferStatusModal";
import { appConstants } from "../../../common";
import { isEqual } from "lodash";
import { marketingActions, newOfferActions } from "../actions";
import { propertyCartActions } from "../../browseSupply/actions";
import PropertiesListItem from "./PropertiesListItem";
import ActiveOffersPagination from "./ActiveOffersPagination";
import styles from "./css/ActiveArchiveOffers.module.less";
import { history } from "../../../helpers";

export default function ActiveOffers() {
  const dispatch = useDispatch();
  const [confirmOfferStatus, setConfirmOfferStatus] = useState({});
  let { offers, loading } = useSelector(
    (state) => window.getValue(state, "offerlistsearchandfilters"),
    isEqual
  );
  const hideModal = useCallback((reloadOffers) => {
    setConfirmOfferStatus({});
    if (reloadOffers) {
      searchAndFilterOffer(appConstants.offerListTab.ACTIVE);
    }
  }, []);

  const offerItemMenu = useCallback(
    (row) => (
      <Menu
        expandIcon={<MoreOutlined className={styles.offerItemMenuIc} />}
        data-testid="menu-change-offer-status"
      >
        <Menu.Item
          key="0"
          disabled={checkIfUserHasRole(appConstants.USER_ROLE.ADMIN) && row.status.name === appConstants.OFFER_STATUS.UNPUBLISHED && row.type === appConstants.PROPERTY_TYPE_LIST[0].id}
          data-testid={row.offerId + "menu-icon"}
          onClick={() => {
            setConfirmOfferStatus({
              offerInfo: row,
              actionType:
                row.status.name === appConstants.OFFER_STATUS.UNPUBLISHED
                  ? appConstants.OFFER_STATUS_ACTION.PUBLISH
                  : appConstants.OFFER_STATUS_ACTION.UNPUBLISH,
            });
          }}
        >
          {row.status.name === appConstants.OFFER_STATUS.UNPUBLISHED ? (
            <FormattedMessage id="nemo.publish" />
          ) : row.status.name === appConstants.OFFER_STATUS.PUBLISHED ? (
            <FormattedMessage id="nemo.unpublish" />
          ) : (
            <Fragment />
          )}
        </Menu.Item>
        <Menu.Item
          key="1"
          disabled={row.status.name === appConstants.OFFER_STATUS.PUBLISHED}
          onClick={() => {
            setConfirmOfferStatus({
              offerInfo: row,
              actionType: appConstants.OFFER_STATUS_ACTION.ARCHIVE,
            });
          }}
        >
          <FormattedMessage id="nemo.archive" />
        </Menu.Item>
      </Menu>
    ),
    []
  );
  const columns = useMemo(
    () => [
      {
        render: (row) => {
          const offerId = row.offerId;
          const offerData = row;
          return (
            <Row gutter={[16, 0]} className={styles.rowWithinTD}>
              <Col span={6} className={styles.imageCol}>
                {row.images && row.images.length > 0 ? (
                  <img
                    onError={(ev) => (ev.target.src = defaultImg)}
                    src={row.images[0].url}
                    alt=""
                    data-testid={row.offerId + "-active-image"}
                    className={styles.image}
                  />
                ) : (
                  <Empty
                    data-testid={row.offerId + "-active-empty-image"}
                    image={Empty.PRESENTED_IMAGE_SIMPLE}
                    className={styles["empty-img"]}
                    description={
                      <span>
                        <FormattedMessage id="nemo.noImages" />
                        &nbsp;
                      </span>
                    }
                  />
                )}
              </Col>
              <Col span={11}>
                <Row className={styles.rowMarginTop}>
                  <Col span={24} className={styles.propNameCol}>
                    <Button
                      type="link"
                      className={styles["offer-name"]}
                      onClick={() => history.push(`/offers/view/${offerId}`)}
                    >
                      <strong>
                        {row.name}&nbsp;-&nbsp;{offerId}&nbsp;
                      </strong>
                    </Button>
                    {row.lcn ? (
                      <span
                        className={styles.propertyBox}
                        data-testid={offerId + "-lcn"}
                      >
                        {<FormattedMessage id="nemo.lcn" />}
                      </span>
                    ) : (
                      <Fragment />
                    )}
                  </Col>
                  <Col span={24}>
                    {process.env.REACT_APP_HIDE_CLIENT_AND_STORE ? (
                      <Fragment />
                    ) : (
                      <b>
                        {row.brandName} - {row.storefrontName}
                      </b>
                    )}
                  </Col>
                  <Col span={24}>
                    <Row gutter={1}>
                      <Col span={16}>
                        <b>
                          <FormattedMessage id="nemo.properties" />
                        </b>
                        {row.properties && row.properties.length > 0 ? (
                          <PropertiesListItem
                            row={row}
                            dataTestId={
                              offerId + "-active-offer-item-properties-list"
                            }
                          />
                        ) : (
                          <Fragment>
                            <br />
                            <FormattedMessage id="nemo.noPropertiesPresent" />
                          </Fragment>
                        )}
                      </Col>
                      <Col span={8}>
                        <Row gutter={[0, 10]}>
                          <Col>
                            <b>
                              <FormattedMessage id="nemo.bookingDates" />
                            </b>
                            <div>
                              {row.displayBookingDateRange ? (
                                `${moment(
                                  row.displayBookingDateRange.startDate
                                ).format("L")} - ${moment(
                                  row.displayBookingDateRange.endDate
                                ).format("L")}`
                              ) : (
                                <FormattedMessage id="nemo.n/a" />
                              )}
                            </div>
                          </Col>
                        </Row>
                        <Row>
                          <Col>
                            <b>
                              <FormattedMessage id="nemo.travelDates" />
                            </b>
                            <div>
                              {row.displayTravellingDateRange ? (
                                `${moment(
                                  row.displayTravellingDateRange.startDate
                                ).format("L")} - ${moment(
                                  row.displayTravellingDateRange.endDate
                                ).format("L")}`
                              ) : (
                                <FormattedMessage id="nemo.n/a" />
                              )}
                            </div>
                          </Col>
                        </Row>
                      </Col>
                    </Row>
                  </Col>
                </Row>
              </Col>
              <Col span={3} className={styles.createdDateCol}>
                <Row className={styles.rowMarginTop}>
                  <b>
                    <FormattedMessage id="nemo.created" />
                  </b>
                  &nbsp;
                  {moment(row.createdDate).format("L")}
                </Row>
                <Row gutter={[0, 15]}>
                  <Col>
                    <Fragment />
                  </Col>
                </Row>
                <Row>
                  <Col offset={10}>
                    <Divider type="vertical" className={styles.divider} />
                  </Col>
                </Row>
              </Col>
              <Col span={4}>
                <Row className={styles.rowMarginTop}>
                  <Col span={14}>
                    {row.status.name === appConstants.OFFER_STATUS.PUBLISHED ? (
                      <span className={styles.enableBox}>
                        <FormattedMessage id="nemo.enable" />
                      </span>
                    ) : row.status.name ===
                      appConstants.OFFER_STATUS.UNPUBLISHED ? (
                      <span className={styles.disableBox}>
                        <FormattedMessage id="nemo.disable" />
                      </span>
                    ) : (
                      <span className={styles.disableBox}>
                        <FormattedMessage id="nemo.unknown" />
                      </span>
                    )}
                  </Col>
                  <Col span={6}>
                    {checkIfUserHasRole(appConstants.USER_ROLE.VIEWER) ? (
                      <Fragment />
                    ) : (
                      <Button
                        type="text"
                        className={styles.editIcon}
                        data-testid={offerId + "active-offer-item-edit-btn"}
                        icon={<EditFilled />}
                        onClick={() => {
                          dispatch({
                            type: newOfferActions.NEW_OFFER_UPDATE_REDUX_STATE,
                            payload: { ...offerData },
                          });
                          dispatch({
                            type: marketingActions.SET_MARKETING_LANGUAGES,
                            payload: parseOfferMarketingLanguagesFromOffer(
                              offerData.languages
                            ),
                          });
                          dispatch({
                            type: propertyCartActions.UPDATE_ENTIRE_CART_LIST,
                            payload: offerData.properties,
                          });
                          dispatch({
                            type: marketingActions.UPDATE_OFFER_MARKETING_INFO,
                            payload: parseOfferMarketingFromOffer(offerData),
                          });
                          history.push(`/offers/edit/${offerId}`);
                          window.scrollTo(0, 0);
                        }}
                      ></Button>
                    )}
                  </Col>
                  <Col span={4}>
                    {checkIfUserHasRole(appConstants.USER_ROLE.VIEWER) ? (
                      <Fragment />
                    ) : (
                      <Dropdown overlay={offerItemMenu(row)}>
                        <MoreOutlined
                          className={styles.moreIcon}
                          data-testid={offerId + "-active-offer-item-menu-btn"}
                        />
                      </Dropdown>
                    )}
                  </Col>
                </Row>
                <Row>
                  <Col>
                    <span className={styles.currencyName}>{row.currency}</span>
                  </Col>
                </Row>
              </Col>
            </Row>
          );
        },
      },
    ],
    [offerItemMenu, dispatch]
  );

  offers =
    offers && offers.map((offer, index) => getParsedOffers(offer, index));

  return (
    <Skeleton loading={loading} active>
      <Table
        dataSource={offers}
        columns={columns}
        rowClassName={styles.rowHeight}
        className={styles.offersListTable}
        showHeader={false}
        data-testid="active-offer-list-table"
        pagination={false}
      />

      <ActiveOffersPagination />

      {confirmOfferStatus.offerInfo ? (
        <ChangeOfferStatusModal
          offerInfo={confirmOfferStatus.offerInfo}
          actionType={confirmOfferStatus.actionType}
          hideModal={hideModal}
        />
      ) : (
        <Fragment />
      )}
    </Skeleton>
  );
}
