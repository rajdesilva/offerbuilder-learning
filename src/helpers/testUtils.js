import React from "react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render as rtlRender } from "@testing-library/react";
import { getValue } from "./utility";
import rootReducers from "../redux/rootReducers";
import * as languages from "../langs";
import { IntlProvider } from "react-intl";
const initialReducerState = {};
window.getValue = getValue;

window.google = {
  maps: {
    Marker: class {},
    Map: class {},
    LatLngBounds: class {},
    places: {
      Autocomplete: class {},
      AutocompleteService: class {},
      PlacesServiceStatus: {
        INVALID_REQUEST: "INVALID_REQUEST",
        NOT_FOUND: "NOT_FOUND",
        OK: "OK",
        OVER_QUERY_LIMIT: "OVER_QUERY_LIMIT",
        REQUEST_DENIED: "REQUEST_DENIED",
        UNKNOWN_ERROR: "UNKNOWN_ERROR",
        ZERO_RESULTS: "ZERO_RESULTS",
      },
      PlacesAutocomplete: {
        INVALID_REQUEST: "INVALID_REQUEST",
        NOT_FOUND: "NOT_FOUND",
        OK: "OK",
        OVER_QUERY_LIMIT: "OVER_QUERY_LIMIT",
        REQUEST_DENIED: "REQUEST_DENIED",
        UNKNOWN_ERROR: "UNKNOWN_ERROR",
        ZERO_RESULTS: "ZERO_RESULTS",
      },
    },

    MarkerClusterer: class {},
    Geocoder: class {},
  },
};
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

function render(
  ui,
  {
    initialState = initialReducerState,
    store = createStore(rootReducers, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <IntlProvider locale="en" messages={languages["en"]}>
          {children}
        </IntlProvider>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
