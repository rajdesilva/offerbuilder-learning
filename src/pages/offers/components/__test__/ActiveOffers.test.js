import React from "react";
import { createStore } from "redux";
import { render } from "../../../../helpers/testUtils";
import * as Util from "../../../../helpers/utility/checkIfUserHasRole";
import { MemoryRouter } from "react-router-dom";
import ActiveOffers from "../ActiveOffers";
import { cloneDeep } from "lodash";

window.fetchWrapper = window.fetchWrapper = jest.fn().mockResolvedValue({
  success: true,
});
const activeOffersStore = {
  offerlistsearchandfilters: {
    selectedTab: "active",
    loading: false,
    offers: [
      {
        id: "16118406",
        internalName: "Cypress Offer - Edit",
        status: "UNPUBLISHED",
        currencyCode: "EUR",
        images: [],
        languages: ["EN"],
        deepLinkSettingsInfo: {
          latitude: 48.2081743,
          longitude: 16.3738189,
          includeAllProperties: true,
          adultOccupancy: 1,
          childOccupancy: 0,
          los: 1,
          rollingOffset: 0,
          checkinDate: "2021-01-28",
          checkoutDate: "2021-01-29",
        },
        marketingInfo: [
          {
            title: "JWM Start Hotels",
            lang: "EN",
            shortDescription: "Experience award-winning service",
            description:
              "Experience award-winning service and sophisticated style at JW Marriott Hotel Vienna.",
            termsAndConditions: "Standard terms and conditions",
          },
        ],
        properties: [
          {
            internalId: 1199,
            propertyCode: "DEMO_BANJARE",
            name: "BANJARE & BANJARE STAR HOTEL",
            channel: "DEMO_OFFERBUILDER",
            supplier: "ntp",
            city: "Vienna",
            rating: 4,
            ratingProvider: "",
            remainingCapitalPool: 0,
            lcn: false,
            images: [],
            marketingInfo: [
              {
                lang: "EN",
                shortDescription: "Experience award-winning service",
                description:
                  "Experience award-winning service and sophisticated style at JW Marriott Hotel Vienna.",
              },
            ],
          },
        ],
        targets: [
          {
            id: "idb",
            name: "Internal Demo Brand",
            storefronts: [
              {
                id: "ids2",
                name: "Best Travel",
                suppliers: [
                  {
                    id: "ntp",
                    name: "NTP",
                    channels: [
                      {
                        id: "DEMO_OFFERBUILDER",
                        name: "DEMO_OFFERBUILDER",
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
        createdAt: "2021-01-28T13:30:33.188+0000",
        updatedAt: "2021-01-28T14:01:21.818+0000",
        lcn: false,
        bookingStartDate: null,
        bookingEndDate: null,
        bookingZoneId: null,
        travelStartDate: null,
        travelEndDate: null,
      },
    ],
    pageOffset: 0,
    pageSize: 10,
    totalOffers: 1,
  },
};
describe("ActiveOffers component test", () => {
  test("create ActiveOffers components snapshot", () => {
    const store = createStore(() => ({ ...activeOffersStore }));
    store.dispatch = jest.fn();
    const { container } = render(
      <MemoryRouter>
        <ActiveOffers />
      </MemoryRouter>,
      {
        store,
      }
    );

    expect(container.firstChild).toMatchSnapshot();
  });
  test("ActiveOffers components- click on 3 dots menu offer button", async () => {
    const store = createStore(() => ({ ...activeOffersStore }));
    store.dispatch = jest.fn();
    const checkUserPermission = jest.spyOn(Util, "checkIfUserHasRole");

    checkUserPermission.mockReturnValue(false); // if user is not viewer
    const { getByTestId } = render(
      <MemoryRouter>
        <ActiveOffers />
      </MemoryRouter>,
      {
        store,
      }
    );
    // first item from offers list
    const menuBtn = getByTestId("16118406-active-offer-item-menu-btn");
    expect(menuBtn).toBeInTheDocument();
  });

  test("ActiveOffers components- menu offer button is not present when user has only viewing role", async () => {
    const store = createStore(() => ({ ...activeOffersStore }));
    store.dispatch = jest.fn();
    const checkUserPermission = jest.spyOn(Util, "checkIfUserHasRole");

    checkUserPermission.mockReturnValue(true); // if user is viewer
    const { queryByTestId } = render(
      <MemoryRouter>
        <ActiveOffers />
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(
      queryByTestId("16118406-active-offer-item-menu-btn")
    ).not.toBeInTheDocument();
  });

  test("ActiveOffers components- if image is present for offer then it is displayed with url given", async () => {
    const tempStore = cloneDeep(activeOffersStore);
    tempStore.offerlistsearchandfilters.offers[0].images = [
      {
        id: "user_1616585539025.png",
        url: "https://res.cloudinary.com/seekda-dev/image/upload/v1616585540/offer_builder_dev/user_1616585539025.jpg",
      },
      {
        id: "user_1616577125384.png",
        url: "https://res.cloudinary.com/seekda-dev/image/upload/v1616577125/offer_builder_dev/user_1616577125384.jpg",
      },
      {
        id: "user_1616585486465.png",
        url: "https://res.cloudinary.com/seekda-dev/image/upload/v1616585487/offer_builder_dev/user_1616585486465.jpg",
      },
    ];
    const store = createStore(() => ({ ...tempStore }));
    store.dispatch = jest.fn();
    const checkUserPermission = jest.spyOn(Util, "checkIfUserHasRole");

    checkUserPermission.mockReturnValue(true); // if user is viewer
    const { getByTestId } = render(
      <MemoryRouter>
        <ActiveOffers />
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(getByTestId("16118406-active-image")).toBeInTheDocument();
  });
});
