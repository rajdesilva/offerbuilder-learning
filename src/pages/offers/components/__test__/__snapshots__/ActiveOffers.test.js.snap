// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveOffers component test create ActiveOffers components snapshot 1`] = `
<div
  class="ant-table-wrapper"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <div
        class="ant-table"
        data-testid="active-offer-list-table"
      >
        <div
          class="ant-table-container"
        >
          <div
            class="ant-table-content"
          >
            <table
              style="table-layout: auto;"
            >
              <colgroup />
              <tbody
                class="ant-table-tbody"
              >
                <tr
                  class="ant-table-row ant-table-row-level-0"
                  data-row-key="16118406"
                >
                  <td
                    class="ant-table-cell"
                  >
                    <div
                      class="ant-row"
                      style="margin-left: -8px; margin-right: -8px;"
                    >
                      <div
                        class="ant-col ant-col-6"
                        style="padding-left: 8px; padding-right: 8px;"
                      >
                        <div
                          class="ant-empty ant-empty-normal"
                          data-testid="16118406-active-empty-image"
                        >
                          <div
                            class="ant-empty-image"
                          >
                            <svg
                              class="ant-empty-img-simple"
                              height="41"
                              viewBox="0 0 64 41"
                              width="64"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <g
                                fill="none"
                                fill-rule="evenodd"
                                transform="translate(0 1)"
                              >
                                <ellipse
                                  class="ant-empty-img-simple-ellipse"
                                  cx="32"
                                  cy="33"
                                  rx="32"
                                  ry="7"
                                />
                                <g
                                  class="ant-empty-img-simple-g"
                                  fill-rule="nonzero"
                                >
                                  <path
                                    d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                  />
                                  <path
                                    class="ant-empty-img-simple-path"
                                    d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  />
                                </g>
                              </g>
                            </svg>
                          </div>
                          <p
                            class="ant-empty-description"
                          >
                            <span>
                              No Images
                               
                            </span>
                          </p>
                        </div>
                      </div>
                      <div
                        class="ant-col ant-col-11"
                        style="padding-left: 8px; padding-right: 8px;"
                      >
                        <div
                          class="ant-row"
                        >
                          <div
                            class="ant-col ant-col-24"
                          >
                            <button
                              class="ant-btn ant-btn-link"
                              type="button"
                            >
                              <strong>
                                Cypress Offer - Edit
                                 - 
                                16118406
                                 
                              </strong>
                            </button>
                          </div>
                          <div
                            class="ant-col ant-col-24"
                          >
                            <b>
                              Internal Demo Brand
                               - 
                              Best Travel
                            </b>
                          </div>
                          <div
                            class="ant-col ant-col-24"
                          >
                            <div
                              class="ant-row"
                              style="margin-left: -0.5px; margin-right: -0.5px;"
                            >
                              <div
                                class="ant-col ant-col-16"
                                style="padding-left: 0.5px; padding-right: 0.5px;"
                              >
                                <b>
                                  Properties
                                </b>
                                <div
                                  class="ant-list ant-list-split ant-list-grid"
                                  data-testid="16118406-active-offer-item-properties-list"
                                >
                                  <div
                                    class="ant-spin-nested-loading"
                                  >
                                    <div
                                      class="ant-spin-container"
                                    >
                                      <div
                                        class="ant-row"
                                        style="margin-left: -8px; margin-right: -8px;"
                                      >
                                        <div>
                                          <div
                                            class="ant-col"
                                            style="padding-left: 8px; padding-right: 8px; flex: 1 1 auto;"
                                          >
                                            <div
                                              class="ant-list-item"
                                            >
                                              <div>
                                                BANJARE & BANJARE STAR HOTEL
                                              </div>
                                            </div>
                                          </div>
                                        </div>
                                        <div>
                                          <div
                                            class="ant-col"
                                            style="padding-left: 8px; padding-right: 8px; flex: 1 1 auto;"
                                          >
                                            <div
                                              class="ant-list-item"
                                            >
                                              <div />
                                            </div>
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                              <div
                                class="ant-col ant-col-8"
                                style="padding-left: 0.5px; padding-right: 0.5px;"
                              >
                                <div
                                  class="ant-row"
                                  style="margin-top: -5px; margin-bottom: 5px;"
                                >
                                  <div
                                    class="ant-col"
                                    style="padding-top: 5px; padding-bottom: 5px;"
                                  >
                                    <b>
                                      Booking dates
                                    </b>
                                    <div>
                                      NA
                                    </div>
                                  </div>
                                </div>
                                <div
                                  class="ant-row"
                                >
                                  <div
                                    class="ant-col"
                                  >
                                    <b>
                                      Travel Dates
                                    </b>
                                    <div>
                                      NA
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="ant-col ant-col-3"
                        style="padding-left: 8px; padding-right: 8px;"
                      >
                        <div
                          class="ant-row"
                        >
                          <b>
                            Created
                          </b>
                           
                          01/28/2021
                        </div>
                        <div
                          class="ant-row"
                          style="margin-top: -7.5px; margin-bottom: 7.5px;"
                        >
                          <div
                            class="ant-col"
                            style="padding-top: 7.5px; padding-bottom: 7.5px;"
                          />
                        </div>
                        <div
                          class="ant-row"
                        >
                          <div
                            class="ant-col ant-col-offset-10"
                          >
                            <div
                              class="ant-divider ant-divider-vertical"
                              role="separator"
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="ant-col ant-col-4"
                        style="padding-left: 8px; padding-right: 8px;"
                      >
                        <div
                          class="ant-row"
                        >
                          <div
                            class="ant-col ant-col-14"
                          >
                            <span>
                              Unpublished
                            </span>
                          </div>
                          <div
                            class="ant-col ant-col-6"
                          >
                            <button
                              class="ant-btn ant-btn-text ant-btn-icon-only"
                              data-testid="16118406active-offer-item-edit-btn"
                              type="button"
                            >
                              <span
                                aria-label="edit"
                                class="anticon anticon-edit"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="edit"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32zm-622.3-84c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9z"
                                  />
                                </svg>
                              </span>
                            </button>
                          </div>
                          <div
                            class="ant-col ant-col-4"
                          >
                            <span
                              aria-label="more"
                              class="anticon anticon-more ant-dropdown-trigger"
                              data-testid="16118406-active-offer-item-menu-btn"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="more"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                                />
                              </svg>
                            </span>
                          </div>
                        </div>
                        <div
                          class="ant-row"
                        >
                          <div
                            class="ant-col"
                          >
                            <span>
                              EUR
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
